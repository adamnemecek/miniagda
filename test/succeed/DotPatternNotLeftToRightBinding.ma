-- 2010-09-22
-- 2012-01-22 parameters gone from constructors

fun A : Set {}
fun B : Set {}
fun f : A -> A {}

data Fix *(a : A) : A -> Set
{ fix : Fix a (f a)
}

-- eta not definable unconditionally (like for Id)
fun eta : (a, b : A) -> Fix a b -> Fix a b
{ eta a .(f a) (fix) = fix
} 

-- variable a used in dot pattern left of its binding
fun bla : (b, a : A) -> Fix a b -> A
{ bla .(f a) a (fix) = a
} 

-- Function inverse

data Inv (g : A -> B) : B -> Set
{ mkInv : (getInv : A) -> Inv g (g getInv)
}
-- MiniAgda does not generate destructor getInv

fun getInv : (g : A -> B) -> (b : B) -> Inv g b -> A
  { getInv g .(g a) (mkInv a) = a 
  }

{- Analysis:

  mkInv : (getInv : A) -> Inv g b  where b = (g getInv)

bind b in destructor type after parameters

  getInv : (g : A -> B) -> (b : B) -> Inv g b -> A

put its value (g a) down as dot-pattern instead of b

  getInv g .(g a) (mkInv .g a) = a

-}

{-
data Id (A : Set)(a : A) : A -> Set
{ refl : Id A a a
}

fun f : (A : Set) -> (c : A -> A) -> (a : A) -> (b : A) -> Id A (c b) b -> A
{ f A c .(c b) b (refl .A .b) = b
}
-}