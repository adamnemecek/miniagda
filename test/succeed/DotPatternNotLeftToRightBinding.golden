MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "DotPatternNotLeftToRightBinding.ma" ---
--- scope checking ---
--- type checking ---
type  A : Set
{}
type  B : Set
{}
none  f : A -> A
{}
type  Fix : (a : A) -> ^ A -> Set
term  Fix.fix : .[a : A] -> < Fix.fix : Fix a (f a) >
term  eta : (a : A) -> (b : A) -> Fix a b -> Fix a b
{ eta a .(f a) Fix.fix = Fix.fix
}
none  bla : (b : A) -> (a : A) -> Fix a b -> A
{ bla .(f a) a Fix.fix = a
}
type  Inv : ^(g : A -> B) -> ^ B -> Set
term  Inv.mkInv : .[g : A -> B] -> ^(getInv : A) -> < Inv.mkInv getInv : Inv g (g getInv) >
none  getInv : (g : A -> B) -> (b : B) -> Inv g b -> A
{ getInv g .(g a) (Inv.mkInv a) = a
}
--- evaluating ---
--- closing "DotPatternNotLeftToRightBinding.ma" ---
